google/api_core/rest_streaming.py

```diff
@@ -0,0 +1,99 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Helpers for server-side streaming in REST."""
+
+import json
+import string
+
+import requests
+
+
+class ResponseIterator:
+    """Iterator over REST API responses.
+
+    Args:
+        response (requests.Response): An API response object.
+        response_message_cls (Callable[proto.Message]): A proto
+        class expected to be returned from an API.
+    """
+
+    def __init__(self, response: requests.Response, response_message_cls):
+        self._response = response
+        self._response_message_cls = response_message_cls
+        # Inner iterator over HTTP response's content.
+        self._response_itr = self._response.iter_content(decode_unicode=True)
+        # Contains a list of JSON responses ready to be sent to user.
+        self._ready_objs = []
+        # Current JSON response being built.
+        self._obj = ""
+        # Keeps track of the nesting level within a JSON object.
+        self._level = 0
+        # Keeps track whether HTTP response is currently sending values
+        # inside of a string value.
+        self._in_string = False
+
+    def cancel(self):
+        """Cancel existing streaming operation.
+        """
+        self._response.close()
+
+    def _process_chunk(self, chunk: str):
+        if self._level == 0:
+            if chunk[0] != "[":
+                raise ValueError(
+                    "Can only parse array of JSON objects, instead got %s" % chunk
+                )
+        for char in chunk:
+            if char == "{":
+                if self._level == 1:
+                    # Level 1 corresponds to the outermost JSON object
+                    # (i.e. the one we care about).
+                    self._obj = ""
+                if not self._in_string:
+                    self._level += 1
+                self._obj += char
+            elif char == '"':
+                self._in_string = not self._in_string
+                self._obj += char
+            elif char == "}":
+                self._obj += char
+                if not self._in_string:
+                    self._level -= 1
+                if not self._in_string and self._level == 1:
+                    self._ready_objs.append(self._obj)
+            elif char in string.whitespace:
+                if self._in_string:
+                    self._obj += char
+            elif char == "[":
+                self._level += 1
+            elif char == "]":
+                self._level -= 1
+            else:
+                self._obj += char
```

Comment:
Looks like it is the most common case, but it comes  last in the whole if-else huge chain, which means almost for every character in the stream we will have to go through every if-else block before reaching the final destination. It is not a very big deal these days, but still feels a bit inefficient. 

Optional: Python does not have proper switch stateements to deal with it, but we can do the a dict instead of if-else. I.e. `if char in whitespace else if char not in special_chars else self._obj += char`, where `special_chars = "{}\"[]"`.

Reply:
Is performance your main concern? I'm not sure if reordering of if-statements will make a significant difference to end user because network I/O is involved. If you insist, I can try reducing if-statements with dict as you suggested.

Reply:
My main concern is readability: if performance becomes a concern we can do some profiling.

Reply:
Yeah, I agree the for-loop is complicated, but it's the least worst option available. I considered other alternatives:

1) Use polymorphism. This removes if-else statements but requires a lot more code and perhaps not so Pythonic here.
2) Create DFA class. This for-loop simulates a finite automata, so I can create a separate FA class and this will hide most of the if-else complexity. But, it requires quite a bit of careful formulation and thought to correctly create states and transitions.
3) What Vadym suggested above (e.g. `dict`). I don't think this improves readability as I either need to nest if-statements or create private functions, which IMO introduces indirection here.

Out of 3 options, the 2nd (DFA) is the best IMO, but it would take me more time to complete, putting our deadlines at risk.